File: frontend/types/rag.ts
export type RagDocumentType = "markdown" | "text" | "json" | "csv";

export interface RagRetrievalOptions {
  dataset?: string;
  document_type?: RagDocumentType;
  top_k?: number;
  minimum_similarity?: number;
}

export interface RagChatRequest extends RagRetrievalOptions {
  question: string;
  stream?: boolean;
}

export interface RagChatResponse {
  answer: string;
  sources: string[];
}

export interface RagSearchRequest extends RagRetrievalOptions {
  query: string;
}

export interface RagSourceMetadata {
  source: string;
  section_title: string;
  dataset_name: string | null;
  document_type: RagDocumentType;
  timestamp: string;
}

export interface RagSearchResult {
  content: string;
  similarity: number;
  metadata: RagSourceMetadata;
}

export interface RagSearchResponse {
  query: string;
  results: RagSearchResult[];
}

export interface RagDocument {
  source: string;
  dataset_name: string | null;
  document_type: RagDocumentType;
  timestamp: string;
  checksum: string;
  chunk_count: number;
}

export interface RagDocumentListResponse {
  items: RagDocument[];
  document_count: number;
  chunk_count: number;
}

export type RagStreamEvent =
  | { type: "token"; content: string }
  | { type: "sources"; sources: string[] }
  | { type: "done" };

File: backend/app/schemas/rag.py
from typing import Literal

from pydantic import BaseModel, Field, model_validator


class RetrievalRequest(BaseModel):
    dataset: str | None = Field(default=None, max_length=200)
    document_type: Literal["markdown", "text", "json", "csv"] | None = None
    top_k: int = Field(default=5, ge=1, le=20)
    minimum_similarity: float | None = Field(default=None, ge=-1, le=1)


class ChatRequest(RetrievalRequest):
    question: str = Field(min_length=2, max_length=2000)
    stream: bool = False

    @model_validator(mode="after")
    def question_is_not_blank(self) -> "ChatRequest":
        if not self.question.strip():
            raise ValueError("Question cannot be blank")
        return self


class SearchRequest(RetrievalRequest):
    query: str = Field(min_length=2, max_length=2000)


class SourceMetadataResponse(BaseModel):
    source: str
    section_title: str
    dataset_name: str | None
    document_type: str
    timestamp: str


class SearchResultResponse(BaseModel):
    content: str
    similarity: float
    metadata: SourceMetadataResponse


class SearchResponse(BaseModel):
    query: str
    results: list[SearchResultResponse]


class ChatResponse(BaseModel):
    answer: str
    sources: list[str]


class DocumentResponse(BaseModel):
    source: str
    dataset_name: str | None
    document_type: str
    timestamp: str
    checksum: str
    chunk_count: int


class DocumentListResponse(BaseModel):
    items: list[DocumentResponse]
    document_count: int
    chunk_count: int


class ReindexRequest(BaseModel):
    force: bool = False


class ReindexResponse(BaseModel):
    document_count: int
    chunk_count: int
    embedding_dimension: int
    embedded_chunks: int
    reused_chunks: int
